external-crates/move/move-compiler/src/parser/syntax.rs

```diff
@@ -2183,6 +2176,55 @@ fn parse_postfix_ability_declarations(
     Ok(())
 }
 
+fn check_struct_visibility(visibility: Option<Visibility>, context: &mut Context) {
+    let current_package = context.package_name;
+    match visibility {
+        Some(Visibility::Public(loc))
+            if context.env.check_feature(
+                &FeatureGate::PublicStructTypeVisibility,
+                current_package,
+                loc,
+            ) => {}
+        Some(vis)
+            if context
+                .env
+                .supports_feature(current_package, &FeatureGate::PublicStructTypeVisibility) =>
+        {
+            let msg = format!(
+                "Invalid struct declaration. Structs can only have have the '{}' visibility modifier",
+                Visibility::PUBLIC
+            );
+            context
+                .env
+                .add_diag(diag!(Syntax::InvalidModifier, (vis.loc().unwrap(), msg)));
```

Comment:
I think we should add a note to all of these diagnostics explaining the breaking change. Thoughts?

Maybe something like `"Starting in Move 2024, visibility must be annotated on struct declarations."

Reply:
Similarly, we should have something for the other cases like `"Visibility is annotated on structs in Move 2024 edition and beyond"`